### PR TITLE
Os 181 :  Adding unique constraints to Neo4j database provider

### PR DESCRIPTION
**Pull request by pritha-tarento**
Adding unique constraints to Neo4j database provider, constraints are added to create Index 
Unique index is kept for higher precedence than index(single, composite)
Resolved parent vertex property update:- by using graph traversal, hence moved the updateParentIndexProperty() to VertexWriter. Called in entityParenter with seperate transaction.

### CODE REVIEW

File: java/registry/src/main/java/io/opensaber/registry/dao/VertexWriter.java [hunk 1]
```diff
@@ -70,6 +70,49 @@ public Vertex createVertex(String label) {
...
+     */
+    private void replaceVertexProperty(String parentlabel, List<String> indexFields, boolean isUnique){
+        String propertyName = isUnique ? Constants.UNIQUE_INDEX_FIELDS : Constants.INDEX_FIELDS;
+        StringBuilder properties = new StringBuilder(String.join(",", indexFields));        
+        P<String> lblPredicate = P.eq(parentlabel);
+        GraphTraversalSource gtRootTraversal = graph.traversal().clone();
+        Iterator<Vertex> iterVertex = gtRootTraversal.V().hasLabel(lblPredicate);
+        if (iterVertex.hasNext()) {
+            Vertex pVertex = iterVertex.next();
+            pVertex.property(propertyName, properties.toString());
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/dao/VertexWriter.java]: Why is the parentVertex can't be reused? The entity parenter was supposed to have a list of uuids with vertices cached, doesn't it?

File: java/registry/src/main/java/io/opensaber/registry/util/EntityParenter.java [hunk 2]
```diff
@@ -212,70 +201,98 @@ public void ensureIndexExists(){
...
-                        status = indexer.createIndex(graph, definition.getTitle(), parentVertex);
-                        logger.info("CREATE INDEX STATUS ====== " + status);
-                        if(status){
-                            indexHelper.updateParentIndexProperty(parentVertex, indexFields, indexUniqueFields);
-                        }
-                                  
+
+    private void asyncAddIndex(DatabaseProvider dbProvider, String shardId, Vertex parentVertex,
+            Definition definition) {
+        // new Thread(() -> {
```
Comment (indrajra) [inline java/registry/src/main/java/io/opensaber/registry/util/EntityParenter.java]: Why is this now sync?

File: java/registry/src/main/resources/public/_schemas/Teacher.json [hunk 3]
```diff
@@ -148,7 +148,7 @@
                       "uniqueIndexFields: Optional; list of field names used for creating unique index. Field names must be different from index field name"],
```
Comment (indrajra) [inline java/registry/src/main/resources/public/_schemas/Teacher.json]: indexFields comment - can you update how to specify a composite index?

File: java/registry/src/main/java/io/opensaber/registry/dao/VertexWriter.java (hunk 1 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/dao/VertexWriter.java]: The update to graph node should happen by a traversal source.

File: java/registry/src/main/java/io/opensaber/registry/util/EntityParenter.java (hunk 2 above)
Comment (pritha-tarento) [reply java/registry/src/main/java/io/opensaber/registry/util/EntityParenter.java]: Remove the comment. Done

File: java/registry/src/main/resources/public/_schemas/Teacher.json (hunk 3 above)
Comment (pritha-tarento) [reply java/registry/src/main/resources/public/_schemas/Teacher.json]: Done.

**Pull request merged**